### PR TITLE
year zero in 360 day valid

### PR DESCRIPTION
**Pull request by marqh**
The year zero special case is for gregorian calendar only

### CODE REVIEW

File: lib/iris/fileformats/pp_rules.py [hunk 1]
```diff
@@ -593,17 +593,17 @@ def date2hours(t):
...
         coords_and_dims.append((DimCoord(t2_epoch_hours, standard_name='forecast_reference_time', units=epoch_hours_unit), None))
 
     # Don't add time coordinates when years are zero and
-    # lbtim.ib == 2. These are handled elsewhere.
+    # lbtim.ic == 2. These are handled elsewhere.
     if \
             (lbtim.ib == 2) and \
-            (lbtim.ic in [1, 2, 4]) and \
-            (t1.year != 0 and t2.year != 0) and \
+            ((lbtim.ic in [1]) and (t1.year != 0 and t2.year != 0) or (lbtim.ic in [2, 4])) and \
```
Comment (pp-mo) [inline lib/iris/fileformats/pp_rules.py]: "(lbtim.ic in [ic])" doesn't make much sense -- it should be (lbtim.ic == 1)

**Comment by pp-mo** [comment]
I'm thinking that this new rules logic should also be reflected in the "vectorised" rules function, ```convert_time_coords```.

**Background**...
The "vectorised" functions are ```_convert_time_coords``` and ```_convert_vertical_coords```,
which parallel the "normal" (scalar, i.e. single-field) functions ```_convert_scalar_time_coords``` and ```_convert_scalar_vertical_coords```.
They are used only for the "fast structured fieldsfile loading", in iris.experimental.fieldsfile.

This insprired me to go looking for changes since the vectorised functions were split from the scalar ones.
I found that there have been a couple of changes to the scalar rules that did not get duplicated in the vector ones :
 * the (lbcode[-1] == 3) case : [here](https://github.com/SciTools/iris/blob/c128574880616e8e84aabfdd46c6c619e5b873c6/lib/iris/fileformats/pp_rules.py#L634)
 * (for the vertical-coordinates) the (lbvc == 19) case : [here](https://github.com/SciTools/iris/blob/c128574880616e8e84aabfdd46c6c619e5b873c6/lib/iris/fileformats/pp_rules.py#L188)

(I think that's all for now, but I could possibly have missed something).

Obviously we are a bit exposed to mistakes here, with the duplicated logic.
Apart from just fixing those, I think it would be desirable to combine the two routines somehow.
There are some problems with that however..
 * the exact logic that 'truly equivalent' vectorised functions would need to replicate includes the behaviour of merge in a 'normal', which is obviously non-trivial
 * @rhattersley pointed out that the vectorised version is currently _intentionally_ different in some ways (i.e. fieldsfile structured loading needs to handle coordinates differently in some cases, at least for now)
   * (most notably the handling of realization coordinate, which structured load conveniently ignores at present)

I am looking into this, and will aim to make a proposal shortly...

**Comment by rhattersley** [comment]
> @rhattersley pointed out that the vectorised version is currently intentionally different in some ways

I didn't mean to imply "intentionally" - it's more a case of "conveniently, in one specific use case(*)". I  wouldn't expect that single use case to prevent the scalar & vector conversions from being made consistent, but the use case is worth bearing in mind when deciding on what that consistent behaviour should be.

*) Data from an ensemble member with index N are written out with LBRSVD4 == N for all time steps except T+0, where they are written with LBRSVD4 == 0. The vectorised conversion ignores this value so multiple time points happily merge. The scalar conversion adds a `realization` coordinate for the non-zero time steps but does nothing at T+0, so they don't merge without manual meddling (e.g. using a load callback).

**Comment by pp-mo** [comment]
> @rhattersley I wouldn't expect that single use case to prevent the scalar & vector conversions from being made consistent, but the use case is worth bearing in mind when deciding on what that consistent behaviour should be.

Understood.

I  think in any case we don't _need_ to completely rationalise the scalar + vector behaviour.
The 'vector' usage is, for now at least, still a special case with it's own independent requirements, which does not need to be identical to 'ordinary' load in every case or respect.

The point at issue here is really the difficulty of maintenance, which can addressed simply by combining the logic into a single routine.
I think we can do that **without** any immediate functional impacts --  then, afterwards, we can discuss what further changes are desirable, to make the two usages more similar.

**Comment by rhattersley** [comment]
> The point at issue here is really the difficulty of maintenance, which can addressed simply by combining the logic into a single routine.

:+1: for reducing the maintenance burden. Care will be needed to ensure unification doesn't significantly worsen performance.

**Comment by pp-mo** [comment]
> this new rules logic should also be reflected in the "vectorised" rules function

Since #1959 merged this should now be easy, but needs a rebase.

**Comment by pelson** [comment]
👍 in principle. @pp-mo, do you want to take over on this PR?

**Comment by pp-mo** [comment]
> @pelson in principle. pp-mo, do you want to take over on this PR?

Ok, probably worth linking with #2243 -- there are a bunch of outstanding issues and to-dos in this area.

**Comment by pelson** [comment]
Closing as this appears to have stalled.

> @pp-mo wrote: Ok, probably worth linking with #2243 -- there are a bunch of outstanding issues and to-dos in this area.

May I strongly encourage small, easily mergeable PRs. This PR wasn't too far away - apart from addressing review comments, adding a simple test and a what's new entry, I'd have been happy to merge.

**Pull request closed without merging**